snap/validate.go

```diff
@@ -71,6 +73,92 @@ func ValidateAlias(alias string) error {
 	return nil
 }
 
+// ValidateAppSocketName checks if a string ca be used as a name for a socket
+// (for socket activation).
+func ValidateAppSocketName(name string) error {
+	valid := validSnapName.MatchString(name)
+	if !valid {
+		return fmt.Errorf("invalid socket name: %q", name)
+	}
+	return nil
+}
+
+// ValidateAppSocketListenAddress checks that the value of socket addresses.
+func ValidateAppSocketListenAddress(socket *SocketInfo, fieldName string, address string) error {
+	if address == "" {
+		return fmt.Errorf("socket %q must define %q", socket.Name, fieldName)
+	}
+
+	switch address[0] {
+	case '/', '$':
+		return validateAppSocketListenAddressPath(socket, fieldName, address)
+	case '@':
+		return validateAppSocketListenAddressAbstractSocket(socket, fieldName, address)
+	default:
+		return validateAppSocketListenAddressNetAddress(socket, fieldName, address)
+	}
+}
+
+func validateAppSocketListenAddressPath(socket *SocketInfo, fieldName string, path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf(
+			`socket %q has invalid %q: paths must not include "." or ".."`, socket.Name, fieldName)
+	}
+
+	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/")) {
+		return fmt.Errorf(
+			"socket %q has invalid %q: only $SNAP_DATA and $SNAP_COMMON prefixes are allowed", socket.Name, fieldName)
+	}
+
+	return nil
+}
+
+func validateAppSocketListenAddressAbstractSocket(socket *SocketInfo, fieldName string, path string) error {
+	prefix := "@snap." + socket.App.Snap.Name()
+	if !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("socket %q path for %q must be prefixed with %q", socket.Name, fieldName, prefix)
```

Comment:
Ah, yes, the prefix has a missing `.` at the end. I'll fix that.

As for the prefix itself, I'd say it's the same reasoning as for the unix path prefix. The user has to know the actual path anyway, since the point of socket activation is to expose your service, so consumers will have to know the path.
Also, for someone reading the snap[craft].yaml, the partial name would be confusing if he doesn't know how it's internally prefixed.

Reply:
Indeed, plus we don't know if we'll want to keep those restrictions forever. Explicit sounds better here.